BootloaderCorePkg/Stage2/Stage2Hob.c

```diff
@@ -605,6 +605,142 @@ BuildUniversalPayloadHob (
+    SmmInfoHob = (LDR_SMM_INFO *)GetGuidHobData (NULL, NULL, &gSmmInformationGuid);
+    ASSERT (SmmInfoHob != NULL);
+
+    Reg = &SmmRegisterHob->Registers[0];
+    Reg->Id                        = REGISTER_ID_SMI_GBL_EN;
+    Reg->Value                     = 1;
+    Reg->Address.AddressSpaceId    = (SmmInfoHob->SmiCtrlReg.RegType == REG_TYPE_IO)?EFI_ACPI_3_0_SYSTEM_IO:EFI_ACPI_3_0_SYSTEM_MEMORY;
+    Reg->Address.RegisterBitWidth  = 1;
+    Reg->Address.RegisterBitOffset = SmmInfoHob->SmiCtrlReg.SmiGblPos;
+    Reg->Address.AccessSize        = EFI_ACPI_3_0_DWORD;
+    Reg->Address.Address           = SmmInfoHob->SmiCtrlReg.Address;
+
+    Reg++;
+    Reg->Id                        = REGISTER_ID_SMI_GBL_EN_LOCK;
+    Reg->Value                     = 1;
+    Reg->Address.AddressSpaceId    = (SmmInfoHob->SmiLockReg.RegType == REG_TYPE_IO)?EFI_ACPI_3_0_SYSTEM_IO:EFI_ACPI_3_0_SYSTEM_MEMORY;
+    Reg->Address.RegisterBitWidth  = 1;
+    Reg->Address.RegisterBitOffset = SmmInfoHob->SmiLockReg.SmiLockPos;
+    Reg->Address.AccessSize        = EFI_ACPI_3_0_DWORD;
+    Reg->Address.Address           = SmmInfoHob->SmiLockReg.Address;
+
+    Reg++;
+    Reg->Id                        = REGISTER_ID_SMI_EOS;
+    Reg->Value                     = 1;
+    Reg->Address.AddressSpaceId    = (SmmInfoHob->SmiCtrlReg.RegType == REG_TYPE_IO)?EFI_ACPI_3_0_SYSTEM_IO:EFI_ACPI_3_0_SYSTEM_MEMORY;
+    Reg->Address.RegisterBitWidth  = 1;
+    Reg->Address.RegisterBitOffset = SmmInfoHob->SmiCtrlReg.SmiEosPos ;
+    Reg->Address.AccessSize        = EFI_ACPI_3_0_DWORD;
+    Reg->Address.Address           = SmmInfoHob->SmiCtrlReg.Address;
+
+    Reg++;
+    Reg->Id                        = REGISTER_ID_SMI_APM_EN;
+    Reg->Value                     = 1;
+    Reg->Address.AddressSpaceId    = (SmmInfoHob->SmiCtrlReg.RegType == REG_TYPE_IO)?EFI_ACPI_3_0_SYSTEM_IO:EFI_ACPI_3_0_SYSTEM_MEMORY;
+    Reg->Address.RegisterBitWidth  = 1;
+    Reg->Address.RegisterBitOffset = SmmInfoHob->SmiCtrlReg.SmiApmPos;
+    Reg->Address.AccessSize        = EFI_ACPI_3_0_DWORD;
+    Reg->Address.Address           = SmmInfoHob->SmiCtrlReg.Address;
+
+    Reg++;
+    Reg->Id                        = REGISTER_ID_SMI_APM_STS;
+    Reg->Value                     = 1;
+    Reg->Address.AddressSpaceId    = (SmmInfoHob->SmiStsReg.RegType == REG_TYPE_IO)?EFI_ACPI_3_0_SYSTEM_IO:EFI_ACPI_3_0_SYSTEM_MEMORY;
+    Reg->Address.RegisterBitWidth  = 1;
+    Reg->Address.RegisterBitOffset = SmmInfoHob->SmiStsReg.SmiApmPos;
+    Reg->Address.AccessSize        = EFI_ACPI_3_0_DWORD;
+    Reg->Address.Address           = SmmInfoHob->SmiStsReg.Address;
+  }
+
+  // SMM memory HOB
+  Length = sizeof (EFI_SMRAM_HOB_DESCRIPTOR_BLOCK) + sizeof (EFI_SMRAM_DESCRIPTOR);
+  SmmMemoryHob = BuildGuidHob (&gEfiSmmSmramMemoryGuid, Length);
+  if (SmmMemoryHob != NULL) {
+    ZeroMem (SmmMemoryHob, Length);
+    SmmInfoHob = (LDR_SMM_INFO *)GetGuidHobData (NULL, NULL, &gSmmInformationGuid);
+    ASSERT (SmmInfoHob != NULL);

```

Comment:
Same comment as  above.

Reply:
The code tried to retrieve gSmmInformationGuid HOB  twice, maybe we just do it once at the beginning instead.

Reply:
Sure. Let me update it.